### PR TITLE
Pull Request for #96: Write LaTeX builder

### PR DESCRIPTION
**Pull request by yuchuan2016**
@stanfordquan @Shun-Yang , can any of you PR when you get time? The executable I use for the builder is `pdflatex`. The builder and the tests are very similar to those of `build_lyx`. I also test it in template, and it seems to work fine.

One thing is that the current builder does not delete auxiliary files like `aux`, etc. The pdf would be compiled much faster after the first run if these auxiliary files exist. We always add these auxiliary files in [.gitignore](https://github.com/gslab-econ/template/blob/eaebce91590df81dcd788c9d8eac000f68d8007e/.gitignore), so they won't get committed.

### CODE REVIEW

**Comment by yuchuan2016** [comment]
Hold off the PR for a moment. I notice one auxiliary file it creates is a log file. I think this should be deleted.

**Comment by yuchuan2016** [comment]
OK. Now the builder will delete the automatically generated log file.

File: gslab_scons/builders/build_latex.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        log_ext = ''
+    log_file = os.path.join(target_dir, ('sconscript%s.log' % log_ext))
+
+    # System call
+    try:
+        command = 'pdflatex -jobname %s %s > %s' % (target_file.replace('.pdf', ''), source_file, log_file)
+        subprocess.check_output(command,
+                                stderr = subprocess.STDOUT,
+                                shell  = True)
+        os.remove(target_file.replace('.pdf', '.log'))
```
Comment (stanfordquan) [inline gslab_scons/builders/build_latex.py]: @yuchuan2016 are we sure we want to delete the log file? We may want to append it into `sconscript.log` before doing so. In the past we've had difficulties debugging tex output.

File: setup.py [hunk 2]
```diff
@@ -57,7 +57,7 @@ def run(self):
 requirements = ['requests']
 
 setup(name         = 'GSLab_Tools',
-      version      = '4.0.1',
+      version      = '4.0.2',
```
Comment (stanfordquan) [inline setup.py]: you're jumping the gun here! we're releasing this for 4.0.1 💃

Comment (Shun-Yang) [inline setup.py]: I thought we are releasing 4.1.0

Comment (yuchuan2016) [inline setup.py]: What I plan is to move step by step from 4.0.1, 4.0.2 to the final 4.1.0...

**Comment by yuchuan2016** [comment]
@stanfordquan , are you running the single test script `test_build_latex.py`? I got the same error when trying to run the single test script. However, I get errors when running every single script (`test_build_lyx`, `test_build_r`, etc). I think we should run the test at the root of the repo by running `python setup.py test`. 

But it's a very good point that `pdflatex` would get stuck and wait for user input when the source file does not exist. I would take a look to see how to solve this.

**Comment by yuchuan2016** [comment]
The reason why I'm reluctant of using `latexmk` as suggested in https://github.com/gslab-econ/template/issues/56#issuecomment-318543908 is that it seems to depend on the installation of Perl on Windows according to [this](http://mg.readthedocs.io/latexmk.html). So I'm not sure it comes with latex/MikTex.

File: gslab_scons/builders/build_latex.py (hunk 1 above)
Comment (yuchuan2016) [inline gslab_scons/builders/build_latex.py]: I think most relevant information is also printed to `sconsript.log`. Below is the comparison for one failed run. Do you think it's still useful to append the `text.log` after `sconscript.log`?

<details> <summary> sconscript.log </summary>
This is pdfTeX, Version 3.14159265-2.6-1.40.17 (TeX Live 2016) (preloaded format=pdflatex)
 restricted \write18 enabled.
entering extended mode
(./source/paper/text.tex
LaTeX2e <2016/03/31>
Babel <3.9r> and hyphenation patterns for 83 language(s) loaded.
(/usr/local/texlive/2016/texmf-dist/tex/latex/base/article.cls
Document Class: article 2014/09/29 v1.4h Standard LaTeX document class
(/usr/local/texlive/2016/texmf-dist/tex/latex/base/size12.clo))

! LaTeX Error: Missing \begin{document}.

See the LaTeX manual or LaTeX Companion for explanation.
Type  H <return>  for immediate help.
 ...                                              
                                                  
l.5 \author{\noindent Notes}\maketitle
                                      
[1{/usr/local/texlive/2016/texmf-var/fonts/map/pdftex/updmap/pdftex.map}]
(./build/paper/text.aux) )
(see the transcript file for additional information)</usr/local/texlive/2016/te
xmf-dist/fonts/type1/public/amsfonts/cm/cmbx12.pfb></usr/local/texlive/2016/tex
mf-dist/fonts/type1/public/amsfonts/cm/cmr12.pfb>
Output written on build/paper/text.pdf (1 page, 18345 bytes).
Transcript written on build/paper/text.log.
 </details>

<details> <summary> text.log </summary>
This is pdfTeX, Version 3.14159265-2.6-1.40.17 (TeX Live 2016) (preloaded format=pdflatex 2016.5.22)  23 AUG 2017 10:57
entering extended mode
 restricted \write18 enabled.
 %&-line parsing enabled.
**source/paper/text.tex
(./source/paper/text.tex
LaTeX2e <2016/03/31>
Babel <3.9r> and hyphenation patterns for 83 language(s) loaded.
(/usr/local/texlive/2016/texmf-dist/tex/latex/base/article.cls
Document Class: article 2014/09/29 v1.4h Standard LaTeX document class
(/usr/local/texlive/2016/texmf-dist/tex/latex/base/size12.clo
File: size12.clo 2014/09/29 v1.4h Standard LaTeX file (size option)
)
\c@part=\count79
\c@section=\count80
\c@subsection=\count81
\c@subsubsection=\count82
\c@paragraph=\count83
\c@subparagraph=\count84
\c@figure=\count85
\c@table=\count86
\abovecaptionskip=\skip41
\belowcaptionskip=\skip42
\bibindent=\dimen102
)

! LaTeX Error: Missing \begin{document}.

See the LaTeX manual or LaTeX Companion for explanation.
Type  H <return>  for immediate help.
 ...                                              
                                                  
l.5 \author{\noindent Notes}\maketitle
                                      
You're in trouble here.  Try typing  <return>  to proceed.
If that doesn't work, type  X <return>  to quit.

LaTeX Font Info:    External font `cmex10' loaded for size
(Font)              <14.4> on input line 5.
LaTeX Font Info:    External font `cmex10' loaded for size
(Font)              <7> on input line 5.
[1

{/usr/local/texlive/2016/texmf-var/fonts/map/pdftex/updmap/pdftex.map}]
(./build/paper/text.aux) ) 
Here is how much of TeX's memory you used:
 214 strings out of 493014
 2390 string characters out of 6133351
 53601 words of memory out of 5000000
 3838 multiletter control sequences out of 15000+600000
 5457 words of font info for 20 fonts, out of 8000000 for 9000
 1141 hyphenation exceptions out of 8191
 22i,6n,19p,123b,187s stack positions out of 5000i,500n,10000p,200000b,80000s
</usr/local/texlive/2016/texmf-dist/fonts/type1/publi
c/amsfonts/cm/cmbx12.pfb></usr/local/texlive/2016/texmf-dist/fonts/type1/public
/amsfonts/cm/cmr12.pfb>
Output written on build/paper/text.pdf (1 page, 18345 bytes).
PDF statistics:
 16 PDF objects out of 1000 (max. 8388607)
 10 compressed objects within 1 object stream
 0 named destinations out of 1000 (max. 500000)
 1 words of extra memory for PDF output out of 10000 (max. 10000000)

 </details>

**Comment by yuchuan2016** [comment]
@stanfordquan , I have added the `-interaction nonstopmode` option in the latex builder. Then `pdflatex` won't stop and wait for user input if it fails. Can you take a look see how it works now?

File: setup.py (hunk 2 above)
Comment (stanfordquan) [inline setup.py]: oops, you guys are right 🍰

**Comment by stanfordquan** [comment]
@yuchuan2016 , the code is still hanging there for me. What happens if you run the test from within the `tests` folder?

I know the point is not that it may fail, but I want to know if it hangs there indefinitely while failing.

**Comment by stanfordquan** [comment]
Looks good @yuchuan2016 , never mind me. Please pull in.

**Comment by stanfordquan** [comment]
Looks good @yuchuan2016 , never mind me. Please pull in.

**Pull request merged**